core/src/main/scala/org/apache/spark/deploy/history/SimpleFsHistoryProvider.scala

```diff
@@ -0,0 +1,308 @@
+import org.apache.spark.scheduler.ReplayListenerBus._
+import org.apache.spark.ui.SparkUI
+import org.apache.spark.util.{Clock, SystemClock, ThreadUtils, Utils}
+
+/**
+ * A class that provides application history from event logs stored in the file system.
+ * This provider will render the history application UI of specifically requested
+ * applications only.  No scanning or caching of event logs from a filesystem is performed.
+ *
+ */
+
+private[history] class SimpleFsHistoryProvider(conf: SparkConf)
+  extends ApplicationHistoryProvider with Logging {
+
+  import FsHistoryProvider._
+
+  private val logDir = conf.getOption("spark.history.fs.logDirectory")
+    .getOrElse(DEFAULT_LOG_DIR)
+
+  private val HISTORY_UI_ACLS_ENABLE = conf.getBoolean("spark.history.ui.acls.enable", false)
+  private val HISTORY_UI_ADMIN_ACLS = conf.get("spark.history.ui.admin.acls", "")
+  private val HISTORY_UI_ADMIN_ACLS_GROUPS = conf.get("spark.history.ui.admin.acls.groups", "")
+  logInfo(s"History server ui acls " + (if (HISTORY_UI_ACLS_ENABLE) "enabled" else "disabled") +
+    "; users with admin permissions: " + HISTORY_UI_ADMIN_ACLS.toString +
+    "; groups with admin permissions" + HISTORY_UI_ADMIN_ACLS_GROUPS.toString)
+
+  private val hadoopConf = SparkHadoopUtil.get.newConfiguration(conf)
+  private val fs = new Path(logDir).getFileSystem(hadoopConf)
+
+  private[history] def initialize(): Unit = {}
+  
+  override def getEventLogsUnderProcess(): Int = 0
+
+  override def getLastUpdatedTime(): Long = 0
+  
+  private val APPL_START_EVENT_PREFIX = "{\"Event\":\"SparkListenerApplicationStart\""
+  private val APPL_END_EVENT_PREFIX = "{\"Event\":\"SparkListenerApplicationEnd\""
+  
+  //Return an empty list iterator
+  override def getListing(): Iterator[FsApplicationHistoryInfo] = List().iterator
+  
+  def getApplicationAttemptInfo(appFileStatus: FileStatus): Option[org.apache.spark.deploy.history.FsApplicationAttemptInfo] = {
+ 
+    // When getting the Application Info, we really only want the key details - Name, ID, start time, stop time
+    val eventsFilter: ReplayEventsFilter = { eventString =>
+      eventString.startsWith(APPL_START_EVENT_PREFIX) ||
+      eventString.startsWith(APPL_END_EVENT_PREFIX)
+    }
+     
+    val appListener = replay(appFileStatus, isApplicationCompleted(appFileStatus), new ReplayListenerBus(), eventsFilter)
+    appListener.appId match {
+      case Some(appId) => {
+        val attemptInfo = new FsApplicationAttemptInfo(
+          appFileStatus.getPath().getName(),
+          appListener.appName.getOrElse("<Not Started>"),
+          appId,
+          appListener.appAttemptId,
+          appListener.startTime.getOrElse(-1L),
+          appListener.endTime.getOrElse(-1L),
+          0L,
+          appListener.sparkUser.getOrElse("<Not Started>"),
+          isApplicationCompleted(appFileStatus),
+          appFileStatus.getLen()
+        )
+        Some(attemptInfo)
+      }
+      case None => None
```

Comment:
You can avoid this pattern:

```
match {
  case Some(...) => Some(do something)
  case None => None
}
```

by mapping over the option. `Option.map(f)` will apply `f` to the contents of the `Option` if they are set otherwise it will return `None`.